File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/JMinecraftVersionList.java
package net.kdt.pojavlaunch;

import androidx.annotation.Keep;

import net.kdt.pojavlaunch.value.DependentLibrary;
import net.kdt.pojavlaunch.value.MinecraftClientInfo;

import java.util.Map;

/**
 * マインクラフトのバージョンリストをJSONからデシリアライズするためのクラス。
 */
@Keep
@SuppressWarnings("unused") // 未使用フィールドはすべてJSON構造の一部です
public class JMinecraftVersionList {
    /** 最新バージョンのマップ（release, snapshot） */
    public Map<String, String> latest;
    /** 全バージョンの配列 */
    public Version[] versions;

    /**
     * ファイルプロパティ（id, sha1, url, size）を保持する基底クラス。
     */
    @Keep
    public static class FileProperties {
        public String id, sha1, url;
        public long size;
    }

    /**
     * 個々のバージョン情報を保持するクラス。
     */
    @Keep
    public static class Version extends FileProperties {
        // 1.13以降の引数
        public Arguments arguments;
        /** アセットインデックス */
        public AssetIndex assetIndex;

        public String assets;
        /** ダウンロード情報のマップ */
        public Map<String, MinecraftClientInfo> downloads;
        /** 継承元バージョン */
        public String inheritsFrom;
        /** Javaバージョン情報 */
        public JavaVersionInfo javaVersion;
        /** 依存ライブラリ */
        public DependentLibrary[] libraries;
        /** メインクラス名 */
        public String mainClass;
        /** マインクラフト引数（1.12以前） */
        public String minecraftArguments;
        /** 最小ランチャーバージョン */
        public int minimumLauncherVersion;
        /** リリース日時 */
        public String releaseTime;
        public String time;
        public String type;
    }

    /**
     * Javaバージョン情報。
     */
    @Keep
    public static class JavaVersionInfo {
        public String component;
        public int majorVersion;
        public int version; // LabyMod 4で使用されるパラメータ
    }

    /**
     * 1.13以降の引数構造。
     */
    @Keep
    public static class Arguments {
        public Object[] game;
        public Object[] jvm;

        @Keep
        public static class ArgValue {
            public ArgRules[] rules;
            public String value;

            // TLauncherスタイルの引数
            public String[] values;

            @Keep
            public static class ArgRules {
                public String action;
                public String features;
                public ArgOS os;

                @Keep
                public static class ArgOS {
                    public String name;
                    public String version;
                }
            }
        }
    }

    /**
     * アセットインデックス情報。
     */
    @Keep
    public static class AssetIndex extends FileProperties {
        public long totalSize;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/value/JMinecraftVersionList.kt
package com.arata.yukarilauncher.value

import androidx.annotation.Keep
import com.arata.yukarilauncher.value.DependentLibrary
import com.arata.yukarilauncher.value.MinecraftClientInfo

/**
 * Minecraftバージョン一覧のJSONデータを表現するクラス。
 * Mojangのバージョンマニフェストをデシリアライズするために使用されます。
 * 最新バージョン情報と利用可能なすべてのバージョンのリストを保持します。
 */
@Keep
@Suppress("unused")
class JMinecraftVersionList {
    /** 最新バージョンのマップ（キー: "release", "snapshot"）。 */
    var latest: MutableMap<String, String>? = null
    /** 利用可能なすべてのバージョンの配列。 */
    @JvmField var versions: Array<Version>? = null

    /**
     * ファイルプロパティの基底クラス。
     * バージョンファイルのID、SHA-1ハッシュ、URL、サイズを保持します。
     */
    @Keep
    open class FileProperties {
        /** ファイルの識別子。 */
        @JvmField var id: String? = null
        /** ファイルのSHA-1ハッシュ値。 */
        var sha1: String? = null
        /** ファイルのダウンロードURL。 */
        var url: String? = null
        /** ファイルのサイズ（バイト単位）。 */
        var size: Long = 0
    }

    /**
     * 個別のMinecraftバージョン情報を保持するクラス。
     * FilePropertiesを継承し、バージョン固有の詳細情報を追加で保持します。
     */
    @Keep
    class Version : FileProperties() {
        /** 起動引数の設定。 */
        var arguments: Arguments? = null
        /** アセットインデックスの情報。 */
        var assetIndex: AssetIndex? = null
        /** アセットのバージョン識別子。 */
        var assets: String? = null
        /** ダウンロード情報のマップ。 */
        var downloads: MutableMap<String, MinecraftClientInfo>? = null
        /** 継承元のバージョンID。 */
        var inheritsFrom: String? = null
        /** Javaバージョン情報。 */
        var javaVersion: JavaVersionInfo? = null
        /** 依存ライブラリの配列。 */
        var libraries: Array<DependentLibrary>? = null
        /** メインクラス名。 */
        var mainClass: String? = null
        /** レガシーなMinecraft起動引数。 */
        var minecraftArguments: String? = null
        /** 必要最小限のランチャーバージョン。 */
        var minimumLauncherVersion: Int = 0
        /** リリース日時。 */
        @JvmField var releaseTime: String? = null
        /** 更新日時。 */
        @JvmField var time: String? = null
        /** バージョンタイプ（"release", "snapshot", "old_beta"など）。 */
        @JvmField var type: String? = null
    }

    /**
     * Javaバージョン情報を保持するクラス。
     */
    @Keep
    class JavaVersionInfo {
        /** コンポーネント名。 */
        var component: String? = null
        /** メジャーバージョン番号。 */
        var majorVersion: Int = 0
        /** バージョン番号。 */
        var version: Int = 0
    }

    /**
     * 起動引数の設定を保持するクラス。
     * game引数とjvm引数をそれぞれ管理します。
     */
    @Keep
    class Arguments {
        /** ゲーム起動引数の配列。 */
        var game: Array<Any>? = null
        /** JVM起動引数の配列。 */
        var jvm: Array<Any>? = null

        /**
         * 引数値を保持するクラス。
         * ルールベースの条件付き引数に対応します。
         */
        @Keep
        class ArgValue {
            /** この引数を適用するルールの配列。 */
            var rules: Array<ArgValue.ArgRules>? = null
            /** 引数の値。 */
            var value: String? = null
            /** 複数の値を持つ場合の配列。 */
            var values: Array<String>? = null

            /**
             * 引数の適用ルールを定義するクラス。
             * アクション、機能要件、OS要件を指定します。
             */
            @Keep
            class ArgRules {
                /** ルールのアクション（"allow" または "disallow"）。 */
                var action: String? = null
                /** 必要な機能。 */
                var features: String? = null
                /** OS条件。 */
                var os: ArgRules.ArgOS? = null

                /**
                 * OS条件を定義するクラス。
                 */
                @Keep
                class ArgOS {
                    /** OS名。 */
                    var name: String? = null
                    /** OSバージョンのパターン。 */
                    var version: String? = null
                }
            }
        }
    }

    /**
     * アセットインデックス情報を保持するクラス。
     * FilePropertiesを継承し、合計サイズを追加で保持します。
     */
    @Keep
    class AssetIndex : FileProperties() {
        /** 全アセットの合計サイズ（バイト単位）。 */
        var totalSize: Long = 0
    }
}